net/golioth/settings.c

```diff
@@ -95,6 +96,27 @@ static int send_status_report(struct golioth_client *client,
 	return send_coap_response(client, response_buf, nbytes);
 }
 
+static void save_setting(const char *key, const struct golioth_settings_value *value)
+{
+	int err;
+	char name[64] = {};
+
+	/*
+	 * Use the "gsettings" package in the Zephyr settings subsystem
+	 * to store all settings.
+	 */
+	snprintf(name, sizeof(name), "gsettings/%s", key);
+
+	if (IS_ENABLED(CONFIG_GOLIOTH_SETTINGS_PERSIST)) {
+		LOG_INF("Saving setting: %s", name);
+		err = settings_save_one(name, value, sizeof(struct golioth_settings_value));
```

Comment:
I don't think saving:
```
struct golioth_settings_value {
	enum golioth_settings_value_type type;
	union {
		bool b;
		float f;
		struct {
			const char *ptr; /* not NULL-terminated */
			size_t len;
		} string;
		int64_t i64;
	};
};
```
and restoring it after firmware upgrade (changing firmware build in general) is 100% safe, because of the following reasons:
1. `struct golioth_settings_value` might change over time - we might further extend it, which will make it binary incompatible,
2. there are endianess problems - potentially user might change endianness of between builds for targets that support both little and big endian (very unlikely, but still),
3. I don't know C standard enough to be sure that `enum` will remain the same size even if we add more values to it, `bool` will occupy the same amount of bytes at the same offset, etc.; all those are subject to toolchain upgrade, so if something is undefined by standard, it is potentially a bug to rely on that.

In the end, to be 100% sure that we save something persistently and then can restore it from by different firmware build, then we need to either:
1. be very strict on how we write those settings, e.g. using `__packed` and paying attention to endianness,
2. use a serialization library like CBOR to handle all the complexities of back and forth compatibility between firmware builds.

In the past I have run into the same kind of problem and figured that it is much better to use CBOR for serializing typed settings. This includes the complexity of writing all the code to serialize and deseriallize CBOR, but it also solves all the complexities and corner cases of how binary representation of the same structure will look like (which is why the serialization libraries exist in the first place). CBOR also allows to save `bool` into single byte, whereas with current code it would occupy 8 bytes (in most platforms/cases).

Reply:
Besides, storing string the way it is now is not going to work (just pointer and length are stored, not the real payload).

Reply:
Well, CBOR also adds type information, so you can easily change the type from `int64_t` to `float` between FW releases and when loading settings from persistent storage one can detect that previously it was an `int64_t` and now it is `float`.

>and I think endianness is not a concern, since it's extremely unlikely a user will change endianness between FW releases

Unlikely, but possible :)

IMO if we are going to add the type of variable to stored settings, then added complexity isn't much different from including CBOR as a dependency. But... hard to say without that being implemented. I guess we would need a sample that actually read those settings, as without it users might wonder how to get into those settings (we are getting way beyond original idea of "dump the structure, read the structure" which was relatively straightforward for users).

Reply:
> Well, CBOR also adds type information, so you can easily change the type from int64_t to float between FW releases and when loading settings from persistent storage one can detect that previously it was an int64_t and now it is float.

True. Although I tend to think that's the user's problem to solve.

> Unlikely, but possible

You're right, but I also think that making the interface more complicated for the .1% hypothetical user is not necessarily a good tradeoff to make.

> IMO if we are going to add the type of variable to stored settings, then added complexity isn't much different from including CBOR as a dependency. But... hard to say without that being implemented. I guess we would need a sample that actually read those settings, as without it users might wonder how to get into those settings (we are getting way beyond original idea of "dump the structure, read the structure" which was relatively straightforward for users)

With all of the points you've made in the PR, I think it's probably not worth trying to store the settings in flash at all. There are just too many things that can go wrong, and the value added is not worth it. The user can save settings however they want (and if they want), and we shouldn't jump through a bunch of hoops trying to do things automatically for them, because it's just too awkward and too much code.

So, I think I'm just going to close this PR.

FWIW, in the ESP-IDF SDK we store the settings to flash automatically (the NVS library has functions to set/get int, bool, float, string, so we are just calling those functions), and there's no regard for possible endianness or type changes between FW releases (i.e. we assume the user will solve this problem if it comes up). I believe the type information is stored in flash by the NVS library, but I'd have to check. I'll probably leave that SDK as is, since I don't think it's terribly important for the SDKs to be consistent on this point.

Reply:
>FWIW, in the ESP-IDF SDK we store the settings to flash automatically (the NVS library has functions to set/get int, bool, float, string, so we are just calling those functions), and there's no regard for possible endianness or type changes between FW releases (i.e. we assume the user will solve this problem if it comes up). I believe the type information is stored in flash by the NVS library, but I'd have to check. I'll probably leave that SDK as is, since I don't think it's terribly important for the SDKs to be consistent on this point.

It would be good if Zephyr supported such feature, so we could just utilize that, same as with ESP-IDF. Storing settings without giving users at least an example how to restore them is like half-baked solution IMO, thus not sure about the overall value this gives to users. And overall the place for this feature is rather Zephyr upstream, because storing settings is orthogonal task to cloud connection. There are also many non-volatile storage options with Zephyr, like EEPROM, NVS, Settings, FCB, FS, raw flash access, etc., so it is hard to satisfy majority of users with something that doesn't turn out to be a subsystem ("typed settings").

Reply:
Agreed. I think if Zephyr were to support typed settings, we could probably resurrect this PR. For now, I think I'll just close it.